### PR TITLE
Fire - Mona

### PR DESCRIPTION
**Pull request by MonaRahmani**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. |  on the Work model I wrote the spotlight method. this method sorts all the works based on their number of votes(in descending order). in a tie situation return the one that alphabetically comes first. 
Describe how you approached testing that model method. What edge cases did you come up with? | it can select works with the most votes
What are `session` and `flash`? What is the difference between them? | session is similar to flash, except that data stored there will not go away after the next request-response cycle. session stores information about a user when they log in until they log out, but in flash data added to flash is only persisted through the end of the next request-response cycle, and after that, it's gone.
What was one thing that you gained more clarity on through this assignment? | I have learnt more about relationships and testing.
What is the Heroku URL of your deployed application? | https://mradies-app.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/views/pages/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    <% @top_work = Work.spotlight %>
+    <%= link_to "#{@top_work.title}", work_path(@top_work), class: "spotlight__link-to" %><%= " by #{@top_work.creator}" %>
+  </h2>
+  <p class="spotlight__description"><%= "#{@top_work.votes.count} Votes - #{@top_work.description}" %></p>
+</section>
+
+<section class="top-ten__container">
+  <%= render partial: 'top-ten_table', locals: {category: "movie"} %>
+  <%= render partial: 'top-ten_table', locals: {category: "book"} %>
+  <%= render partial: 'top-ten_table', locals: {category: "album"} %>
```
Comment (beccaelenzil) [inline app/views/pages/index.html.erb:13]: Great use of partial views!

File: app/models/work.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  validates :title, presence: true, uniqueness: true
+  validates :category, presence: true
+  validates :creator, presence: true
+  validates :publication_year, presence: true
+
+  # relationship
+  has_many :votes
+  has_many :users, through: :votes
+
+  def self.spotlight
```
Comment (beccaelenzil) [inline app/models/work.rb:12]: Nice, clear, readable model methods.

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+
+  private
+
+  def work_params
+    return params.require(:work).permit(:category, :title, :creator, :publication_year, :description)
+  end
+
+  def find_work
```
Comment (beccaelenzil) [inline app/controllers/works_controller.rb:76]: Good use of a controller filter

**Review by beccaelenzil — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️ Consider redirecting instead of showing 404.
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔ Styles are greata!
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

Great work overall. It is clear that the learning goals around working with session and flash, creating styles to match an existing site, creating custom routes and methods, and testing using fixtures were met. Your code is clear, readable, and well tested. Keep up the hard work!

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
